actions/kubectl-actions.py

```diff
@@ -0,0 +1,77 @@
+#!/usr/local/sbin/charm-env python3
+import os
+import json
+import tempfile
+import subprocess
+from charmhelpers.core.hookenv import (
+    action_get,
+    action_set,
+    action_fail,
+    action_name
+)
+from charmhelpers.core.templating import render
+
+
+def _kubectl(args):
+    """
+    Executes kubectl with args as arguments
+    """
+    snap_bin = os.path.join(os.sep, "snap", "bin")
+    env = os.environ.copy()
+    env["PATH"] = os.pathsep.join([snap_bin, env["PATH"]])
+    cmd = ["kubectl", "--kubeconfig=/home/ubuntu/config"]
+    cmd.extend(args)
+    return subprocess.check_output(
+        cmd, env=env,
+        stderr=subprocess.STDOUT,
+    )
+
+
+def get_kubeconfig():
+    """
+    Read the kubeconfig on the master and return it as JSON
+    """
+    try:
+        result = _kubectl(["config", "view", "-o", "json", "--raw"])
+        # JSON format verification
+        kubeconfig = json.dumps(json.loads(result))
+        action_set({"kubeconfig": kubeconfig})
+    except json.JSONDecodeError as e:
+        action_fail("Failed to parse kubeconfig: {}".format(str(e)))
+    except Exception as e:
+        action_fail("Failed to retrieve kubeconfig: {}".format(str(e)))
+
+
+def apply_manifest():
+    """
+    Applies a user defined manifest with kubectl
+    """
+    _, apply_path = tempfile.mkstemp(suffix=".json")
+    try:
+        manifest = json.loads(action_get("json"))
```

Comment:
Is the reasoning behind the json.loads -> json.dump to try and catch formatting issues before kubectl apply in the hopes that json.loads would give better error messages?

Reply:
Yeah that is the indeed the reasoning.